### ISSUE TITLE
Large allocations from mempool (primarily isomalloc)

### ISSUE THREAD

**Issue by PhilMiller** [UIUC-PPL/charm#198]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/198*

---
There seems to be a serious problem in the memory allocation scheme provided by mempool under the gemini_gni-crayxe layer for AMPI. A simple example shown here demonstrates the problem, which is arising in actual executions of Jacobi and BRAMS, under gemini_gni-crayxe on Blue Waters and JYC with GNU compilers.

Although the mempool source code (mempool.c, mempool.h) looks to have been designed to support up to 1GB of allocation, the actual limit available is much lower. The code example below, which declares a 1 GB array that is sliced across the number of processors in use, will work fine with 8 processors (i.e. 128 MB allocated by each processor), but when it's run with 4 processors (i.e. 256 MB per processor) it fails on every processor with error-messages like

[0] CmiAbort: Mempool-requested slot is more than what mempool can provide as    one chunk, increase cutOffNum and cutoffPoints in mempool

So, apparently there's something that is wrongly limiting the amount of memory that one can allocate, which would mean a bug. Also, 1GB is probably a low limit these days, anyway, as a Cray-XE node has 64 GB of memory, and even when running one process per integer core, each process has 2GB of memory available -- in fact, using fewer tasks per node, each process can use even more than that. So, once the current bug is fixed, it would also be good to increase that current 1 GB limit, if at all possible.

This test was done with GNU compilers (module PrgEnv-gnu loaded) on JYC, using AMPI from a Charm checkout on March-6/2013. I left the example code shown below, plus a Makefile, job-submission script and execution outputs, on JYC under ~cmendes/BUG-AMPI/BUG-3/

AMPI Code Example:
```
SUBROUTINE MPI_MAIN
include 'mpif.h'

PARAMETER (N=16384)
REAL, ALLOCATABLE :: A(:,:)
INTEGER comm, myrank, p, ierr, m

comm=MPI_COMM_WORLD
CALL MPI_INIT(IERR)
CALL MPI_COMM_SIZE(comm, p, ierr)
CALL MPI_COMM_RANK(comm, myrank, ierr)

m = N/p
ALLOCATE (A(N,m))

CALL MPI_BARRIER(comm, ierr)
print *,"[",myrank,"] Done!"
CALL MPI_FINALIZE(IERR)

END

```

**Comment by nikhil-jain** [comment]
*Original date: 2013-06-07 01:38:49*

---
Patched in a fix to 6.5 that allows allocations till 1 GB (minus the mempool header size). A generalized solution is been added to master for 6.6.

**Comment by PhilMiller** [comment]
*Original date: 2013-08-28 18:55:37*

---
No further complaints from Celso or Alvaro, so presumably this is sufficiently dealt with.
